### ISSUE TITLE
Validation is not working properly in case of oneOf of the schamas

### ISSUE THREAD

**Issue by akanduri** [cdimascio/express-openapi-validator#275]
Validation is not working properly in case of oneOf of the schemas

Here are the details,

Note: few properties are the same name both of the schemas.

Post call :
      oneOf:
          #ref: registerFormObj
          #ref: registrationDetailsFormObj

registerFormObj:
      {
  "register_type": "XXX",
  "register_token": {
    "register_value": "1234",
    "registration_source": "PO",
    "regproc_source": "IU"
      },
  "register_alias": "cumminity",
  "store": true
  }

registrationDetailsFormObj:
      {
  "register_type": "XXX",
  "register_token": {
    "register_value": "1234",
    "registration_source": "PO",
    "regproc_source": "IU"
      },
  "register_alias": "cumminity",
  "store": true,
   "created_time": "2019-03-29",
  "updated_time": "2019-03-29",
  "archived_time": "2019-03-29",
  }

The validation framework has to validation either of the one object based on one input right.
In my case, if I give invalid input of the first Object, but validation framework doing two times validations.

Getting a response like below:

{
  "message": "request.body.tender_type should be equal to one of the allowed values: BC, GC, request.body.register_type should be equal to one of the allowed values: BC, GC, request.body should match exactly one schema in oneOf",
  "errors": [
    {
      "path": ".body.register_type",
      "message": "should be equal to one of the allowed values: BC, GC",
      "errorCode": "enum.openapi.validation"
    },
    {
      "path": ".body.register_type",
      "message": "should be equal to one of the allowed values: BC, GC",
      "errorCode": "enum.openapi.validation"
    },
    {
      "path": ".body",
      "message": "should match exactly one schema in oneOf",
      "errorCode": "oneOf.openapi.validation"
    }
  ]
}

Could you please work on a high priority.

**Comment by cdimascio** [comment]
@akanduri can you provide you openapi spec. i'll take a look

**Comment by theBoEffect** [comment]
@akanduri , sorry if this isn't applicable, but I just wanted to let you know I did run into this recently as well. The problem was that my request matched more than one of the oneOf objects and it must only match 1 of them (exactly one), which looks like the same problem here. You could probably make this work by adding an "additionalProperties: false" to your registerFormObj. Alternatively, since you have not declared that additional properties are not allowed, you don't actually need the oneOf, the registerFormObj spec would suffice.

**Comment by nrail-joerg-walter** [comment]
I am struggling with:

```
    Test:
      oneOf:
        - $ref: '#/components/schemas/Test2'
      type: object
      additionalProperties: false
      properties:
        testProp1:
          type: string
      required:
        - testProp1

    Test2:
      type: object
      additionalProperties: false
      properties:
        testProp2:
          type: string
      required:
        - testProp2
```

with

```
{
"testProp2": "string",
"testProp1": "string"
}
```

this results in :

```
{
    "message": "request.body should NOT have additional properties, request.body should NOT have additional properties, request.body should match exactly one schema in oneOf"
}
```
Is this error correct?

**Comment by cdimascio** [comment]
I believe that is expected. 
tldr, Try removing `additionalProperties: false`

Here is some information with respect to use of `anyOf`, `oneOf` etc with `additionalProperties`

https://github.com/ajv-validator/ajv/blob/master/FAQ.md#additional-properties-inside-compound-keywords-anyof-oneof-etc

**Comment by cdimascio** [comment]
@theBoEffect with regard to you comment:
> The problem was that my request matched more than one of the oneOf objects and it must only match 1 of them (exactly one), which looks like the same problem here.

For objects that may match multiple schemas, you might choose `anyOf` instead of `oneOf`. anyOf will pass validation if at least 1 matches, rather than rather than exactly 1.

**Comment by cdimascio** [comment]
@akanduri there is now support for top-level discriminators which 

see the [discriinator example](https://github.com/cdimascio/express-openapi-validator/tree/master/examples/8-top-level-discriminator)
